ledger-service/http-json/src/main/scala/com/digitalasset/http/DbStartupOps.scala

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.http
+
+import com.daml.http.dbbackend.ContractDao
+import com.daml.logging.{ContextualizedLogger, LoggingContextOf}
+import cats.effect.IO
+import com.daml.http.util.Logging.InstanceUUID
+import scalaz.std.option._
+
+object DbStartupOps {
+
+  private[this] val logger = ContextualizedLogger.get(getClass)
+
+  sealed trait DbVersionState
+  object DbVersionState {
+    final case object Missing extends DbVersionState
+    final case object UpToDate extends DbVersionState
+    final case class Mismatch(expected: Int, actual: Int) extends DbVersionState
+  }
+  import DbStartupMode._
+  import DbVersionState._
+
+  def getDbVersionState(implicit
+      dao: ContractDao,
+      lc: LoggingContextOf[InstanceUUID],
+  ): IO[Option[DbVersionState]] = {
+    import dao.jdbcDriver
+    for {
+      _ <- IO(logger.info("Checking for existing schema in the DB"))
+      currentSchemaVersion = jdbcDriver.queries.schemaVersion
+      getVersionResult <- dao.transact(jdbcDriver.queries.version()).attempt
+    } yield getVersionResult
+      .fold(
```

Comment:
probably for a separate PR: maybe consider returning the error in an `Either` instead of returning `None`. That makes it very clear what the meaning of `Left` is where as `None` can be anything.

Reply:
Deal :) I hesitated to go with an `Either[Unit, _]` :sweat_smile: But I agree that with `Option` is is more confusing right now.

Reply:
`Either[Throwable, _]` should also work I believe.

Reply:
Adjusted that small nit with my last build fix :) Hopefully the build finally passes :smile: